### PR TITLE
MTDSA-26291: FHL Removal | BISS API | Add tax year validation for v2

### PR DESCRIPTION
**Pull request by GazalNebHMRC**

### CODE REVIEW

File: app/v3/retrieveBiss/RetrieveBISSService.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      "INVALID_CORRELATION_ID"    -> InternalError,
+      "INVALID_TAXABLE_ENTITY_ID" -> NinoFormatError,
+      "INVALID_INCOME_SOURCETYPE" -> InternalError,
+      "TAX_YEAR_NOT_SUPPORTED"    -> RuleTaxYearNotSupportedError
+    )
+
+    errors ++ extraTysErrors
+  }
+
+}
```
Comment (iyadi) [inline app/v3/retrieveBiss/RetrieveBISSService.scala:68]: is there any reason for having `errors` and `extraTysErrors` 
could you not do 
```scala
private val downstreamErrorMap: Map[String, MtdError] = Map(
      "INVALID_IDVALUE"              -> NinoFormatError,
      "INVALID_TAXYEAR"              -> TaxYearFormatError,
      "INVALID_IDTYPE"               -> InternalError,
      "INVALID_CORRELATIONID"        -> InternalError,
      "INVALID_INCOMESOURCETYPE"     -> InternalError,
      "INVALID_INCOMESOURCEID"       -> BusinessIdFormatError,
      "INCOME_SUBMISSIONS_NOT_EXIST" -> RuleNoIncomeSubmissionsExist,
      "INVALID_ACCOUNTING_PERIOD"    -> InternalError,
      "INVALID_QUERY_PARAM"          -> InternalError,
      "NOT_FOUND"                    -> NotFoundError,
      "SERVER_ERROR"                 -> InternalError,
      "SERVICE_UNAVAILABLE"          -> InternalError,
      "INVALID_TAX_YEAR"          -> TaxYearFormatError,
      "INVALID_INCOMESOURCE_ID"   -> BusinessIdFormatError,
      "INVALID_CORRELATION_ID"    -> InternalError,
      "INVALID_TAXABLE_ENTITY_ID" -> NinoFormatError,
      "INVALID_INCOME_SOURCETYPE" -> InternalError,
      "TAX_YEAR_NOT_SUPPORTED"    -> RuleTaxYearNotSupportedError
    )
``` 
it is your call
it is just a suggestion

File: test/v3/retrieveBiss/def1/model/response/LossSpec.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+class LossSpec extends UnitSpec {
+
+  private val json = Json.parse("""
+      |{
+      |        "net": 0.00,
+      |        "taxable": 35.00
+      |}
+      |""".stripMargin)
+
```
Comment (iyadi) [inline test/v3/retrieveBiss/def1/model/response/LossSpec.scala:30]: can you please tidy the json

```scala
 private val json = Json.parse("""
      |{
      |        "net": 0.00,
      |        "taxable": 35.00
      |}
      |""".stripMargin)
      ```
      to 
      ```scala
       private val json = Json.parse("""
      |{
      |   "net": 0.00,
      |   "taxable": 35.00
      |}
      |""".stripMargin)
      ```

File: test/v3/retrieveBiss/def1/model/response/ProfitSpec.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+class ProfitSpec extends UnitSpec {
+
+  private val json = Json.parse(
+    """
+      |{
+      |        "net": 0.00,
+      |        "taxable": 35.00
+      |}
+      |""".stripMargin)
```
Comment (iyadi) [inline test/v3/retrieveBiss/def1/model/response/ProfitSpec.scala:30]: json format

File: test/v3/retrieveBiss/def1/model/response/TotalSpec.scala [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    """
+      |{
+      |        "income": 100.00,
+      |        "expenses": 50.00,
+      |        "additions": 5.00,
+      |        "deductions": 60.00,
+      |        "accountingAdjustments": -30.00
+      |}
+      |""".stripMargin)
+
```
Comment (iyadi) [inline test/v3/retrieveBiss/def1/model/response/TotalSpec.scala:34]: json format

File: test/v3/retrieveBiss/RetrieveBISSServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      val extraTysErrors = Seq(
+        "INVALID_TAX_YEAR"          -> TaxYearFormatError,
+        "INVALID_INCOMESOURCE_ID"   -> BusinessIdFormatError,
+        "INVALID_CORRELATION_ID"    -> InternalError,
+        "INVALID_TAXABLE_ENTITY_ID" -> NinoFormatError,
+        "INVALID_INCOME_SOURCETYPE" -> InternalError,
+        "TAX_YEAR_NOT_SUPPORTED"    -> RuleTaxYearNotSupportedError
+      )
+
+      (errors ++ extraTysErrors).foreach(args => (serviceError _).tupled(args))
```
Comment (iyadi) [inline test/v3/retrieveBiss/RetrieveBISSServiceSpec.scala:92]: We can simplify this code 
```scala
val errors = Seq(
        "INVALID_IDVALUE"              -> NinoFormatError,
        "INVALID_TAXYEAR"              -> TaxYearFormatError,
        "INVALID_IDTYPE"               -> InternalError,
        "INVALID_CORRELATIONID"        -> InternalError,
        "INVALID_INCOMESOURCETYPE"     -> InternalError,
        "INVALID_INCOMESOURCEID"       -> BusinessIdFormatError,
        "INCOME_SUBMISSIONS_NOT_EXIST" -> RuleNoIncomeSubmissionsExist,
        "INVALID_ACCOUNTING_PERIOD"    -> InternalError,
        "INVALID_QUERY_PARAM"          -> InternalError,
        "NOT_FOUND"                    -> NotFoundError,
        "SERVER_ERROR"                 -> InternalError,
        "SERVICE_UNAVAILABLE"          -> InternalError
      )

      val extraTysErrors = Seq(
        "INVALID_TAX_YEAR"          -> TaxYearFormatError,
        "INVALID_INCOMESOURCE_ID"   -> BusinessIdFormatError,
        "INVALID_CORRELATION_ID"    -> InternalError,
        "INVALID_TAXABLE_ENTITY_ID" -> NinoFormatError,
        "INVALID_INCOME_SOURCETYPE" -> InternalError,
        "TAX_YEAR_NOT_SUPPORTED"    -> RuleTaxYearNotSupportedError
      )

      (errors ++ extraTysErrors).foreach(args => (serviceError _).tupled(args))
```
to 
```scala
val error = Seq(
  "INVALID_IDVALUE"              -> NinoFormatError,
  "INVALID_TAXYEAR"              -> TaxYearFormatError,
  "INVALID_IDTYPE"               -> InternalError,
  "INVALID_CORRELATIONID"        -> InternalError,
  "INVALID_INCOMESOURCETYPE"     -> InternalError,
  "INVALID_INCOMESOURCEID"       -> BusinessIdFormatError,
  "INCOME_SUBMISSIONS_NOT_EXIST" -> RuleNoIncomeSubmissionsExist,
  "INVALID_ACCOUNTING_PERIOD"    -> InternalError,
  "INVALID_QUERY_PARAM"          -> InternalError,
  "NOT_FOUND"                    -> NotFoundError,
  "SERVER_ERROR"                 -> InternalError,
  "SERVICE_UNAVAILABLE"          -> InternalError,
  "INVALID_TAX_YEAR"             -> TaxYearFormatError,
  "INVALID_INCOMESOURCE_ID"      -> BusinessIdFormatError,
  "INVALID_CORRELATION_ID"       -> InternalError,
  "INVALID_TAXABLE_ENTITY_ID"    -> NinoFormatError,
  "INVALID_INCOME_SOURCETYPE"    -> InternalError,
  "TAX_YEAR_NOT_SUPPORTED"       -> RuleTaxYearNotSupportedError
)

erros.foreach { 
    case (downstreamErrorCode, error) =>  serviceError(downstreamErrorCode, error)
}
```
no need to lift the function to accept tuples 

it is your call
p.s. I can see this pattern have been used a lot 
`xxxx.foreach(args => (serviceError _).tupled(args))`
in my opinion this makes the code more obscure

**Review by iyadi — COMMENTED** [review]
some suggestion to simplify readability

File: app/v3/retrieveBiss/RetrieveBISSService.scala (hunk 1 above)
Comment (GazalNebHMRC) [reply app/v3/retrieveBiss/RetrieveBISSService.scala:68]: It has been done like this in all APIs to distinguish the backend errors we receive from tys and non-tys downstream URLs

**Review by iyadi — APPROVED** [review]

**Pull request merged**